### ISSUE TITLE
ECONNRESET with SSL KeepAlive connections + idle timeout + connection pool

### ISSUE THREAD

**Issue by candlerb** [zatosource/zato#465]
Scenario:

* zato making outbound SSL connections
* Remote host is running Windows IIS (7.5 and 8 tested)
* KeepAlives are active (which is the default at both sides)
* Connection drops after idle timeout

tcpdump shows that the remote IIS host sends a RST on idle timeout. This appears to be IIS-specific behaviour  (Apache2 and haproxy both close the connection normally)

At that point nothing bad happens immediately. However the connection apparently remains in the pool. The next time you send a request to this host, it is reused; some data is sent down it; this then raises an exception. After upgrading pyOpenSSL as per #464, it becomes visible as `OpenSSL.SSL.SysCallError: (104, 'ECONNRESET')`. Prior to this you just see `Error: []`

Full details of how to reproduce this are in [this posting](https://mailman-mail5.webfaction.com/pipermail/zato-discuss/2015-June/001339.html)

It appears to be specific to zato or gevent, since running the requests library _outside_ of zato doesn't show the problem.

Here's an strace of the gunicorn worker when the problem occurs:

~~~
[pid 30455] poll([{fd=57, events=POLLIN}], 1, 0) = 1 ([{fd=57, revents=POLLIN|POLLERR|POLLHUP}])
[pid 30455] gettimeofday({1434982140, 237913}, NULL) = 0
[pid 30455] write(57, "\25\3\1\0 \257\213\f\206K|\315\210-\353MoZ\207BAr&\346\305='^vx\250\306".
.., 37) = -1 ECONNRESET (Connection reset by peer)
~~~

which I believe is from `is_connection_dropped(conn)` in `requests/packages/urllib3/util/connection.py`

~~~
    p = poll()
    p.register(sock, POLLIN)
    for (fno, ev) in p.poll(0.0):
        if fno == sock.fileno():
            # Either data is buffered (bad), or the connection is dropped.
            return True
~~~

which in turn is called from `requests/packages/urllib3/connectionpool.py`

~~~
        if conn and is_connection_dropped(conn):
            log.info("Resetting dropped connection: %s" % self.host)
            conn.close()
~~~

I am guessing that it has detected the underlying socket has closed, but is still trying to send an SSL-level close message down it.

**Comment by dsuch** [comment]
@candlerb How about in 2.0.6 we update the dependencies needed and work around it by catching the exception and re-trying, once only?

If on this one retry it is still an error it means there is really something fishy and we let it propagate?

What say you?

**Comment by candlerb** [comment]
I think that's a reasonable workaround; it's what I'm currently doing in user code, and if the framework did it then it would be invisible.

**Comment by dsuch** [comment]
@candlerb - OK, I understand you recommend upgrading the dependencies per #464 ?

**Comment by candlerb** [comment]
Yes I think so. Otherwise you will get a generic error (OpenSSL.SSL.Error) rather than a SysCallError.

**Comment by gaaf** [comment]
I have a fix in my fork which seems to work fine with the updated dependencies from #464

**Comment by dsuch** [comment]
Done in 45bd24d2972385aadf1c39110be346e162699aec - will be included in 2.0.6. Thanks @candlerb  and @gaaf !

**Comment by candlerb** [comment]
FYI a fix for this has now been [merged upstream](https://github.com/shazow/urllib3/pull/729) in urllib3.
